// @bitauth/libauth/build/main/lib/transaction/transaction-e2e.spec.helper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.hdPublicKey7H = exports.hdPrivateKey7H = exports.hdPublicKey6H = exports.hdPrivateKey6H = exports.hdPublicKey5H = exports.hdPrivateKey5H = exports.hdPublicKey4H = exports.hdPrivateKey4H = exports.hdPublicKey3H = exports.hdPrivateKey3H = exports.hdPublicKey2H = exports.hdPrivateKey2H = exports.hdPublicKey1H = exports.hdPrivateKey1H = exports.hdPublicKey0H = exports.hdPrivateKey0H = exports.hdPublicKey = exports.hdPrivateKey = exports.sigOfSigJson = exports.p2pkhJson = exports.cashChannelsJson = exports.twoOfThreeJson = exports.twoOfTwoRecoverableJson = exports.oneOfEightTreeJson = void 0;
const oneOfEightTreeJson = __importStar(require("./fixtures/templates/1-of-8-tree.json"));
exports.oneOfEightTreeJson = oneOfEightTreeJson;
const twoOfTwoRecoverableJson = __importStar(require("./fixtures/templates/2-of-2-recoverable.json"));
exports.twoOfTwoRecoverableJson = twoOfTwoRecoverableJson;
const twoOfThreeJson = __importStar(require("./fixtures/templates/2-of-3.json"));
exports.twoOfThreeJson = twoOfThreeJson;
const cashChannelsJson = __importStar(require("./fixtures/templates/cash-channels.json"));
exports.cashChannelsJson = cashChannelsJson;
const p2pkhJson = __importStar(require("./fixtures/templates/p2pkh.json"));
exports.p2pkhJson = p2pkhJson;
const sigOfSigJson = __importStar(require("./fixtures/templates/sig-of-sig.json"));
exports.sigOfSigJson = sigOfSigJson;
/**
 * `m` from `zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo zoo wrong`
 */
exports.hdPrivateKey = 'xprv9s21ZrQH143K2PfMvkNViFc1fgumGqBew45JD8SxA59Jc5M66n3diqb92JjvaR61zT9P89Grys12kdtV4EFVo6tMwER7U2hcUmZ9VfMYPLC';
/**
 * `M`
 */
exports.hdPublicKey = 'xpub661MyMwAqRbcEsjq2muW5PYkDikFgHuWJGzu1WrZiQgHUsgEeKMtGducsZe1iRsGAGNGDzmWYDM69ya24LMyR7mDhtzqQsc286XEQfM2kkV';
/**
 * `m/0'`
 */
exports.hdPrivateKey0H = 'xprv9uNAm3qC8EoibXd3mwgQ9rxF8XJdfA9V9sF25DpLtYcf1u51Rpf8tfV4n2PdChXM97miXGiJf6UL2SsathXbiVbF6tSgmAVGM3XNb6Yn2EZ';
/**
 * HD public key for `m/0'`
 */
exports.hdPublicKey0H = 'xpub68MXAZN5xcN1p1hWsyDQWztygZ984csLX6AcscDxSt9dthQ9yMyPSToYdJ24jCS5jaVMGSiLeGuP2cWvgKKYQsNXyg988XGGQYgk1FjDv4P';
/**
 * `m/1'`
 */
exports.hdPrivateKey1H = 'xprv9uNAm3qC8EoicAhBeTEJyZE9DCPtmoQ2noPeFCiDxSqgTGV8ZwM89SAyw8Hhygg3Lkgqy53XM2nHtyMmhyVuDwxJupg9F3bemDN6XcdYNJd';
/**
 * HD public key for `m/1'`
 */
exports.hdPublicKey1H = 'xpub68MXAZN5xcN1pemekUmKLhAsmEEPBG7tA2KF3b7qWnNfL4pH7UfNhEVTnQGeMZycFEGbS7w2gsyG7w77YWkoupkn6QZvRq7zeFS98Fic25b';
/**
 * `m/2'`
 */
exports.hdPrivateKey2H = 'xprv9uNAm3qC8EoieXyZAVbnB5LeeCMKUKe8HcE9mi38ghdBrjgay1ixgt9Ve2EjvB4udeZdp7XeyvU4PyezsjVP6xmjBDcsidobgAYCBLVYyrj';
/**
 * HD public key for `m/2'`
 */
exports.hdPublicKey2H = 'xpub68MXAZN5xcN1s242GX8nYDHPCEBosnMyeq9ka6SkF3AAjY1jWZ3DEgTyVKFwPNXcKsY1snLZV7rNcVdYUpL3eFWTrr8xnHhRW4XGmcgHdr7';
/**
 * `m/3'`
 */
exports.hdPrivateKey3H = 'xprv9uNAm3qC8EoiichvTy7ZtvcWx9ZDbr3C9MC6idYAGQwKjU8Xpf94z3DwYQ74MrZj2rBZvoxT6yWiTsDx1PRSLKbCDjm1w6GcNzeCwU1iX4x';
/**
 * HD public key for `m/3'`
 */
exports.hdPublicKey3H = 'xpub68MXAZN5xcN1w6nPZzeaG4ZFWBPi1Jm3Wa7hX1wmpkUJcGTgNCTKXqYRPgUqfxntGEFWQ28CmtcdMQRk6upkK9JEv4wgWuHrEtVhYBToWzt';
/**
 * `m/4'`
 */
exports.hdPrivateKey4H = 'xprv9uNAm3qC8Eoijbhusk2ytrP34sX2wU6xo1Mh9PWXMbbNRUEKW4659fmHB9J6TdHsVURixeb6ZhjJrHtRTGzMRpLSWeoy9AxJGwonLTHTQVw';
/**
 * HD public key for `m/4'`
 */
exports.hdPublicKey4H = 'xpub68MXAZN5xcN1x5nNymZzFzKmcuMXLvppAEHHwmv8uw8MJGZU3bQKhU5m2PNpJP837U5SymxHLNRPJP7XWXWMbQ6JgkEiYVCdkgp2ANasobg';
/**
 * `m/5'`
 */
exports.hdPrivateKey5H = 'xprv9uNAm3qC8EoipVBzuYyPsmYPxdp2DY3343Wivkr19jgrMcoRkqkJQ63sLG8jx1Za217Dk9x7zgCGipAK2J3NoLW2cVJK8Yg68kPRXCDggVR';
/**
 * HD public key for `m/5'`
 */
exports.hdPublicKey5H = 'xpub68MXAZN5xcN22yGU1aWQEuV8WfeWczktRGSKj9Fci5DqER8aJP4YwtNMBZP2j5jjYimDGsM5uQmyzL8K9mPZJBSr7RqC15MxpApnZyAdjyQ';
/**
 * `m/6'`
 */
exports.hdPrivateKey6H = 'xprv9uNAm3qC8EoirM9LS7M1qs1nQbkVXPKqPYDSGMTghdGUETrYCewDU9B2x87kfcBAmSfCu9WzdXRoqnGCKBYeVWEhzMphFmuxd6QTJn3J747';
/**
 * HD public key for `m/6'`
 */
exports.hdPublicKey6H = 'xpub68MXAZN5xcN24qDoY8t2CzxWxdayvr3gkm934jsJFxoT7GBgkCFU1wVWoNDfYji7GxUxuVGQMMLqYFgnucWetCuUBACxiZZedMm92ZUrVYT';
/**
 * `m/7'`
 */
exports.hdPrivateKey7H = 'xprv9uNAm3qC8EoisrNfAA4SDj4gHSYt6Y2ziVtWm61c66D8d7ZxQacYuJEbjcu3bVWVCnahooFPJ1HZ3QBpvoxzP8enJ71GwN2s3m4mqco5zbx';
/**
 * HD public key for `m/7'`
 */
exports.hdPublicKey7H = 'xpub68MXAZN5xcN26LT8GBbSas1QqUPNVzkr5ip7ZURDeRk7Vuu6x7voT6Z5atLN1n6sxRgMbtV1tNaHVn4iAtbWJBHRgdeDn3tZKRqSpfNx7X5';
