// @bitauth/libauth/build/main/lib/vm/instruction-sets/bch/bch-instruction-sets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createInstructionSetBCH = exports.getFlagsForInstructionSetBCH = exports.instructionSetBCHCurrentStrict = exports.InstructionSetBCH = exports.isWitnessProgram = exports.isPayToScriptHash = exports.OpcodesBCH = void 0;
const combinators_1 = require("../common/combinators");
const common_1 = require("../common/common");
const instruction_sets_utils_1 = require("../instruction-sets-utils");
const bch_errors_1 = require("./bch-errors");
const bch_opcodes_1 = require("./bch-opcodes");
Object.defineProperty(exports, "OpcodesBCH", { enumerable: true, get: function () { return bch_opcodes_1.OpcodesBCH; } });
const bch_operations_1 = require("./bch-operations");
exports.isPayToScriptHash = (verificationInstructions) => verificationInstructions.length === 3 /* length */ &&
    verificationInstructions[0].opcode ===
        bch_opcodes_1.OpcodesBCH.OP_HASH160 &&
    verificationInstructions[1].opcode ===
        bch_opcodes_1.OpcodesBCH.OP_PUSHBYTES_20 &&
    verificationInstructions[2 /* lastElement */]
        .opcode === bch_opcodes_1.OpcodesBCH.OP_EQUAL;
/**
 * Test a stack item for the SegWit Recovery Rules activated in `BCH_2019_05`.
 *
 * @param bytecode - the stack item to test
 */
// eslint-disable-next-line complexity
exports.isWitnessProgram = (bytecode) => {
    const correctLength = bytecode.length >= 4 /* minimumLength */ &&
        bytecode.length <= 42 /* maximumLength */;
    const validVersionPush = bytecode[0] === 0 /* OP_0 */ ||
        (bytecode[0] >= 81 /* OP_1 */ && bytecode[0] <= 96 /* OP_16 */);
    const correctLengthByte = bytecode[1] + 2 /* versionAndLengthBytes */ === bytecode.length;
    return correctLength && validVersionPush && correctLengthByte;
};
/**
 * From C++ implementation:
 * Note that IsPushOnly() *does* consider OP_RESERVED to be a push-type
 * opcode, however execution of OP_RESERVED fails, so it's not relevant to
 * P2SH/BIP62 as the scriptSig would fail prior to the P2SH special
 * validation code being executed.
 */
const isPushOperation = (opcode) => opcode < bch_opcodes_1.OpcodesBCH.OP_16;
/**
 * This library's supported versions of the BCH virtual machine. "Strict"
 * versions (A.K.A. `isStandard` from the C++ implementations) enable additional
 * validation which is commonly used on the P2P network before relaying
 * transactions. Transactions which fail these rules are often called
 * "non-standard" – the transactions can technically be included by miners in
 * valid blocks, but most network nodes will refuse to relay them.
 *
 * BCH instruction sets marked `SPEC` ("specification") have not yet been
 * deployed on the main network and are subject to change. After deployment, the
 * `SPEC` suffix is removed. This change only effects the name of the TypeScript
 * enum member – the value remains the same. E.g.
 * `InstructionSetBCH.BCH_2020_05_SPEC` became `InstructionSetBCH.BCH_2020_05`,
 * but the value remained `BCH_2020_05`.
 *
 * This allows consumers to select an upgrade policy: when a version of Libauth
 * is released in which compatibility with a deployed virtual machine is
 * confirmed, this change can help to identify downstream code which requires
 * review.
 *  - Consumers which prefer to upgrade manually should specify a `SPEC` type,
 * e.g. `InstructionSetBCH.BCH_2020_05_SPEC`.
 *  - Consumers which prefer full compatibility between Libauth version should
 * specify a precise instruction set value (e.g. `BCH_2020_05`) or use the
 * dedicated "current" value: `instructionSetBCHCurrentStrict`.
 */
var InstructionSetBCH;
(function (InstructionSetBCH) {
    InstructionSetBCH["BCH_2019_05"] = "BCH_2019_05";
    InstructionSetBCH["BCH_2019_05_STRICT"] = "BCH_2019_05_STRICT";
    InstructionSetBCH["BCH_2019_11"] = "BCH_2019_11";
    InstructionSetBCH["BCH_2019_11_STRICT"] = "BCH_2019_11_STRICT";
    InstructionSetBCH["BCH_2020_05"] = "BCH_2020_05";
    InstructionSetBCH["BCH_2020_05_STRICT"] = "BCH_2020_05_STRICT";
    InstructionSetBCH["BCH_2020_11_SPEC"] = "BCH_2020_11";
    InstructionSetBCH["BCH_2020_11_STRICT_SPEC"] = "BCH_2020_11_STRICT";
    InstructionSetBCH["BCH_2021_05_SPEC"] = "BCH_2021_05";
    InstructionSetBCH["BCH_2021_05_STRICT_SPEC"] = "BCH_2021_05_STRICT";
    InstructionSetBCH["BCH_2021_11_SPEC"] = "BCH_2021_11";
    InstructionSetBCH["BCH_2021_11_STRICT_SPEC"] = "BCH_2021_11_STRICT";
    InstructionSetBCH["BCH_2022_05_SPEC"] = "BCH_2022_05";
    InstructionSetBCH["BCH_2022_05_STRICT_SPEC"] = "BCH_2022_05_STRICT";
    InstructionSetBCH["BCH_2022_11_SPEC"] = "BCH_2022_11";
    InstructionSetBCH["BCH_2022_11_STRICT_SPEC"] = "BCH_2022_11_STRICT";
})(InstructionSetBCH = exports.InstructionSetBCH || (exports.InstructionSetBCH = {}));
/**
 * The current strict virtual machine version used by the Bitcoin Cash (BCH)
 * network.
 */
exports.instructionSetBCHCurrentStrict = InstructionSetBCH.BCH_2020_05_STRICT;
// eslint-disable-next-line complexity
exports.getFlagsForInstructionSetBCH = (instructionSet) => {
    switch (instructionSet) {
        case InstructionSetBCH.BCH_2019_05:
            return {
                disallowUpgradableNops: false,
                opReverseBytes: false,
                requireBugValueZero: false,
                requireMinimalEncoding: false,
                requireNullSignatureFailures: true,
            };
        case InstructionSetBCH.BCH_2019_05_STRICT:
            return {
                disallowUpgradableNops: true,
                opReverseBytes: false,
                requireBugValueZero: false,
                requireMinimalEncoding: true,
                requireNullSignatureFailures: true,
            };
        case InstructionSetBCH.BCH_2019_11:
            return {
                disallowUpgradableNops: false,
                opReverseBytes: false,
                requireBugValueZero: true,
                requireMinimalEncoding: true,
                requireNullSignatureFailures: true,
            };
        case InstructionSetBCH.BCH_2019_11_STRICT:
            return {
                disallowUpgradableNops: true,
                opReverseBytes: false,
                requireBugValueZero: true,
                requireMinimalEncoding: true,
                requireNullSignatureFailures: true,
            };
        case InstructionSetBCH.BCH_2020_05:
            return {
                disallowUpgradableNops: false,
                opReverseBytes: true,
                requireBugValueZero: false,
                requireMinimalEncoding: false,
                requireNullSignatureFailures: true,
            };
        case InstructionSetBCH.BCH_2020_05_STRICT:
            return {
                disallowUpgradableNops: true,
                opReverseBytes: true,
                requireBugValueZero: true,
                requireMinimalEncoding: true,
                requireNullSignatureFailures: true,
            };
        default:
            return new Error(`${instructionSet} is not a known instruction set.`);
    }
};
/**
 * Initialize a new instruction set for the BCH virtual machine.
 *
 * @param flags - an object configuring the flags for this vm (see
 * `getFlagsForInstructionSetBCH`)
 * @param sha1 - a Sha1 implementation
 * @param sha256 - a Sha256 implementation
 * @param ripemd160 - a Ripemd160 implementation
 * @param secp256k1 - a Secp256k1 implementation
 */
exports.createInstructionSetBCH = ({ flags, ripemd160, secp256k1, sha1, sha256, }) => (Object.assign(Object.assign({ clone: common_1.cloneAuthenticationProgramStateCommon, continue: (state) => state.error === undefined && state.ip < state.instructions.length, 
    // eslint-disable-next-line complexity
    evaluate: (program, stateEvaluate) => {
        var _a;
        const { unlockingBytecode } = program.spendingTransaction.inputs[program.inputIndex];
        const { lockingBytecode } = program.sourceOutput;
        const unlockingInstructions = instruction_sets_utils_1.parseBytecode(unlockingBytecode);
        const lockingInstructions = instruction_sets_utils_1.parseBytecode(lockingBytecode);
        const externalState = common_1.createTransactionContextCommon(program);
        const initialState = common_1.createAuthenticationProgramStateCommon({
            instructions: unlockingInstructions,
            stack: [],
            transactionContext: externalState,
        });
        const unlockingResult = unlockingBytecode.length > common_1.ConsensusCommon.maximumBytecodeLength
            ? common_1.applyError(common_1.AuthenticationErrorCommon.exceededMaximumBytecodeLengthUnlocking, initialState)
            : instruction_sets_utils_1.authenticationInstructionsAreMalformed(unlockingInstructions)
                ? common_1.applyError(common_1.AuthenticationErrorCommon.malformedUnlockingBytecode, initialState)
                : lockingBytecode.length > common_1.ConsensusCommon.maximumBytecodeLength
                    ? common_1.applyError(common_1.AuthenticationErrorCommon.exceededMaximumBytecodeLengthLocking, initialState)
                    : instruction_sets_utils_1.authenticationInstructionsAreMalformed(lockingInstructions)
                        ? common_1.applyError(common_1.AuthenticationErrorCommon.malformedLockingBytecode, initialState)
                        : initialState.instructions.every((instruction) => isPushOperation(instruction.opcode))
                            ? stateEvaluate(initialState)
                            : common_1.applyError(bch_errors_1.AuthenticationErrorBCH.requiresPushOnly, initialState);
        if (unlockingResult.error !== undefined) {
            return unlockingResult;
        }
        const lockingResult = stateEvaluate(common_1.createAuthenticationProgramStateCommon({
            instructions: lockingInstructions,
            stack: unlockingResult.stack,
            transactionContext: externalState,
        }));
        if (!exports.isPayToScriptHash(lockingInstructions)) {
            return lockingResult;
        }
        const p2shStack = common_1.cloneStack(unlockingResult.stack);
        // eslint-disable-next-line functional/immutable-data
        const p2shScript = (_a = p2shStack.pop()) !== null && _a !== void 0 ? _a : Uint8Array.of();
        if (p2shStack.length === 0 && exports.isWitnessProgram(p2shScript)) {
            return lockingResult;
        }
        const p2shInstructions = instruction_sets_utils_1.parseBytecode(p2shScript);
        return instruction_sets_utils_1.authenticationInstructionsAreMalformed(p2shInstructions)
            ? Object.assign(Object.assign({}, lockingResult), { error: bch_errors_1.AuthenticationErrorBCH.malformedP2shBytecode }) : stateEvaluate(common_1.createAuthenticationProgramStateCommon({
            instructions: p2shInstructions,
            stack: p2shStack,
            transactionContext: externalState,
        }));
    }, operations: Object.assign(Object.assign({}, common_1.commonOperations({ flags, ripemd160, secp256k1, sha1, sha256 })), combinators_1.mapOverOperations(bch_operations_1.bitcoinCashOperations({
        flags,
        secp256k1,
        sha256,
    }), combinators_1.conditionallyEvaluate, combinators_1.incrementOperationCount, common_1.checkLimitsCommon)) }, common_1.undefinedOperation()), { verify: (state) => {
        if (state.error !== undefined) {
            return state.error;
        }
        if (state.executionStack.length !== 0) {
            return common_1.AuthenticationErrorCommon.nonEmptyExecutionStack;
        }
        if (state.stack.length !== 1) {
            return common_1.AuthenticationErrorCommon.requiresCleanStack;
        }
        if (!common_1.stackItemIsTruthy(state.stack[0])) {
            return common_1.AuthenticationErrorCommon.unsuccessfulEvaluation;
        }
        return true;
    } }));
