// @bitauth/libauth/build/module/lib/template/language/resolve.js
import { hexToBin, utf8ToBin } from '../../format/format';
import { bigIntToScriptNumber } from '../../vm/instruction-sets/instruction-sets';
import { compileScriptRaw } from './compile';
import { IdentifierResolutionErrorType, IdentifierResolutionType, } from './language-types';
import { stringifyErrors } from './language-utils';
const pluckRange = (node) => ({
    endColumn: node.end.column,
    endLineNumber: node.end.line,
    startColumn: node.start.column,
    startLineNumber: node.start.line,
});
const removeNumericSeparators = (numericLiteral) => numericLiteral.replace(/_/gu, '');
export const resolveScriptSegment = (segment, resolveIdentifiers) => {
    // eslint-disable-next-line complexity
    const resolved = segment.value.map((child) => {
        const range = pluckRange(child);
        switch (child.name) {
            case 'Identifier': {
                const identifier = child.value;
                const result = resolveIdentifiers(identifier);
                const ret = result.status
                    ? {
                        range,
                        type: 'bytecode',
                        value: result.bytecode,
                        ...(result.type === IdentifierResolutionType.opcode
                            ? {
                                opcode: identifier,
                            }
                            : result.type === IdentifierResolutionType.variable
                                ? {
                                    ...('debug' in result ? { debug: result.debug } : {}),
                                    ...('signature' in result
                                        ? { signature: result.signature }
                                        : {}),
                                    variable: identifier,
                                }
                                : // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
                                    result.type === IdentifierResolutionType.script
                                        ? { script: identifier, source: result.source }
                                        : { unknown: identifier }),
                    }
                    : {
                        ...('debug' in result ? { debug: result.debug } : {}),
                        ...('recoverable' in result && result.recoverable
                            ? {
                                missingIdentifier: identifier,
                                owningEntity: result.entityOwnership,
                            }
                            : {}),
                        range,
                        type: 'error',
                        value: result.error,
                    };
                return ret;
            }
            case 'Push':
                return {
                    range,
                    type: 'push',
                    value: resolveScriptSegment(child.value, resolveIdentifiers),
                };
            case 'Evaluation':
                return {
                    range,
                    type: 'evaluation',
                    value: resolveScriptSegment(child.value, resolveIdentifiers),
                };
            case 'BigIntLiteral':
                return {
                    literal: child.value,
                    literalType: 'BigIntLiteral',
                    range,
                    type: 'bytecode',
                    value: bigIntToScriptNumber(BigInt(removeNumericSeparators(child.value))),
                };
            case 'BinaryLiteral':
                return {
                    literal: child.value,
                    literalType: 'BinaryLiteral',
                    range,
                    type: 'bytecode',
                    value: bigIntToScriptNumber(BigInt(`0b${removeNumericSeparators(child.value)}`)),
                };
            case 'HexLiteral':
                return {
                    literal: child.value,
                    literalType: 'HexLiteral',
                    range,
                    type: 'bytecode',
                    value: hexToBin(removeNumericSeparators(child.value)),
                };
            case 'UTF8Literal':
                return {
                    literal: child.value,
                    literalType: 'UTF8Literal',
                    range,
                    type: 'bytecode',
                    value: utf8ToBin(child.value),
                };
            case 'Comment':
                return {
                    range,
                    type: 'comment',
                    value: child.value,
                };
            default:
                return {
                    range,
                    type: 'error',
                    value: `Unrecognized segment: ${child.name}`,
                };
        }
    });
    return resolved.length === 0
        ? [{ range: pluckRange(segment), type: 'comment', value: '' }]
        : resolved;
};
export var BuiltInVariables;
(function (BuiltInVariables) {
    BuiltInVariables["currentBlockTime"] = "current_block_time";
    BuiltInVariables["currentBlockHeight"] = "current_block_height";
    BuiltInVariables["signingSerialization"] = "signing_serialization";
})(BuiltInVariables || (BuiltInVariables = {}));
const attemptCompilerOperation = ({ data, environment, identifier, matchingOperations, operationExample = 'operation_identifier', operationId, variableId, variableType, }) => {
    if (matchingOperations === undefined) {
        return {
            error: `The "${variableId}" variable type can not be resolved because the "${variableType}" operation has not been included in this compiler's CompilationEnvironment.`,
            status: 'error',
        };
    }
    if (typeof matchingOperations === 'function') {
        const operation = matchingOperations;
        return operation(identifier, data, environment);
    }
    if (operationId === undefined) {
        return {
            error: `This "${variableId}" variable could not be resolved because this compiler's "${variableType}" operations require an operation identifier, e.g. '${variableId}.${operationExample}'.`,
            status: 'error',
        };
    }
    const operation = matchingOperations[operationId];
    if (operation === undefined) {
        return {
            error: `The identifier "${identifier}" could not be resolved because the "${variableId}.${operationId}" operation is not available to this compiler.`,
            status: 'error',
        };
    }
    return operation(identifier, data, environment);
};
/**
 * If the identifer can be successfully resolved as a variable, the result is
 * returned as a Uint8Array. If the identifier references a known variable, but
 * an error occurs in resolving it, the error is returned as a string.
 * Otherwise, the identifier is not recognized as a variable, and this method
 * simply returns `false`.
 *
 * @param identifier - The full identifier used to describe this operation, e.g.
 * `owner.signature.all_outputs`.
 * @param data - The `CompilationData` provided to the compiler
 * @param environment - The `CompilationEnvironment` provided to the compiler
 */
export const resolveVariableIdentifier = ({ data, environment, identifier, }) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const [variableId, operationId] = identifier.split('.');
    switch (variableId) {
        case BuiltInVariables.currentBlockHeight:
            return attemptCompilerOperation({
                data,
                environment,
                identifier,
                matchingOperations: (_a = environment.operations) === null || _a === void 0 ? void 0 : _a.currentBlockHeight,
                operationId,
                variableId,
                variableType: 'currentBlockHeight',
            });
        case BuiltInVariables.currentBlockTime:
            return attemptCompilerOperation({
                data,
                environment,
                identifier,
                matchingOperations: (_b = environment.operations) === null || _b === void 0 ? void 0 : _b.currentBlockTime,
                operationId,
                variableId,
                variableType: 'currentBlockTime',
            });
        case BuiltInVariables.signingSerialization:
            return attemptCompilerOperation({
                data,
                environment,
                identifier,
                matchingOperations: (_c = environment.operations) === null || _c === void 0 ? void 0 : _c.signingSerialization,
                operationExample: 'version',
                operationId,
                variableId,
                variableType: 'signingSerialization',
            });
        default: {
            const expectedVariable = (_d = environment.variables) === null || _d === void 0 ? void 0 : _d[variableId];
            if (expectedVariable === undefined) {
                return { status: 'skip' };
            }
            return attemptCompilerOperation({
                data,
                environment,
                identifier,
                operationId,
                variableId,
                ...{
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    AddressData: {
                        matchingOperations: (_e = environment.operations) === null || _e === void 0 ? void 0 : _e.addressData,
                        variableType: 'addressData',
                    },
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    HdKey: {
                        matchingOperations: (_f = environment.operations) === null || _f === void 0 ? void 0 : _f.hdKey,
                        operationExample: 'public_key',
                        variableType: 'hdKey',
                    },
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    Key: {
                        matchingOperations: (_g = environment.operations) === null || _g === void 0 ? void 0 : _g.key,
                        operationExample: 'public_key',
                        variableType: 'key',
                    },
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    WalletData: {
                        matchingOperations: (_h = environment.operations) === null || _h === void 0 ? void 0 : _h.walletData,
                        variableType: 'walletData',
                    },
                }[expectedVariable.type],
            });
        }
    }
};
/**
 * Compile an internal script identifier.
 *
 * @remarks
 * If the identifer can be successfully resolved as a script, the script is
 * compiled and returned as a CompilationResultSuccess. If an error occurs in
 * compiling it, the error is returned as a string.
 *
 * Otherwise, the identifier is not recognized as a script, and this method
 * simply returns `false`.
 *
 * @param identifier - the identifier of the script to be resolved
 * @param data - the provided CompilationData
 * @param environment - the provided CompilationEnvironment
 * @param parentIdentifier - the identifier of the script which references the
 * script being resolved (for detecting circular dependencies)
 */
export const resolveScriptIdentifier = ({ data, environment, identifier, }) => {
    if (environment.scripts[identifier] === undefined) {
        return false;
    }
    const result = compileScriptRaw({ data, environment, scriptId: identifier });
    if (result.success) {
        return result;
    }
    return `Compilation error in resolved script "${identifier}": ${stringifyErrors(result.errors)}`;
    /*
     * result.errors.reduce(
     *   (all, { error, range }) =>
     *     `${
     *       all === '' ? '' : `${all}; `
     *     } [${
     *       range.startLineNumber
     *     }, ${range.startColumn}]: ${error}`,
     *   ''
     * );
     */
};
/**
 * Return an `IdentifierResolutionFunction` for use in `resolveScriptSegment`.
 *
 * @param scriptId - the `id` of the script for which the resulting
 * `IdentifierResolutionFunction` will be used.
 * @param environment - a snapshot of the context around `scriptId`. See
 * `CompilationEnvironment` for details.
 * @param data - the actual variable values (private keys, shared wallet data,
 * shared address data, etc.) to use in resolving variables.
 */
export const createIdentifierResolver = ({ data, environment, }) => 
// eslint-disable-next-line complexity
(identifier) => {
    var _a;
    const opcodeResult = (_a = environment.opcodes) === null || _a === void 0 ? void 0 : _a[identifier];
    if (opcodeResult !== undefined) {
        return {
            bytecode: opcodeResult,
            status: true,
            type: IdentifierResolutionType.opcode,
        };
    }
    const variableResult = resolveVariableIdentifier({
        data,
        environment,
        identifier,
    });
    if (variableResult.status !== 'skip') {
        return variableResult.status === 'error'
            ? {
                ...('debug' in variableResult
                    ? { debug: variableResult.debug }
                    : {}),
                error: variableResult.error,
                ...(environment.entityOwnership === undefined
                    ? {}
                    : {
                        entityOwnership: environment.entityOwnership[identifier.split('.')[0]],
                    }),
                recoverable: 'recoverable' in variableResult,
                status: false,
                type: IdentifierResolutionErrorType.variable,
            }
            : {
                ...('debug' in variableResult
                    ? { debug: variableResult.debug }
                    : {}),
                bytecode: variableResult.bytecode,
                ...('signature' in variableResult
                    ? {
                        signature: variableResult.signature,
                    }
                    : {}),
                status: true,
                type: IdentifierResolutionType.variable,
            };
    }
    const scriptResult = resolveScriptIdentifier({
        data,
        environment,
        identifier,
    });
    if (scriptResult !== false) {
        return typeof scriptResult === 'string'
            ? {
                error: scriptResult,
                scriptId: identifier,
                status: false,
                type: IdentifierResolutionErrorType.script,
            }
            : {
                bytecode: scriptResult.bytecode,
                source: scriptResult.resolve,
                status: true,
                type: IdentifierResolutionType.script,
            };
    }
    return {
        error: `Unknown identifier "${identifier}".`,
        status: false,
        type: IdentifierResolutionErrorType.unknown,
    };
};
