// @bitauth/libauth/build/module/lib/crypto/sha1.js
import { instantiateRustWasm, sha1Base64Bytes } from '../bin/bin';
import { base64ToBin } from '../format/format';
/**
 * The most performant way to instantiate sha1 functionality. To avoid
 * using Node.js or DOM-specific APIs, you can use `instantiateSha1`.
 *
 * @param webassemblyBytes - A buffer containing the sha1 binary.
 */
export const instantiateSha1Bytes = async (webassemblyBytes) => {
    const wasm = await instantiateRustWasm(webassemblyBytes, './sha1', 'sha1', 'sha1_init', 'sha1_update', 'sha1_final');
    return {
        final: wasm.final,
        hash: wasm.hash,
        init: wasm.init,
        update: wasm.update,
    };
};
export const getEmbeddedSha1Binary = () => base64ToBin(sha1Base64Bytes).buffer;
const cachedSha1 = {};
/**
 * An ultimately-portable (but slower) version of `instantiateSha1Bytes`
 * which does not require the consumer to provide the sha1 binary buffer.
 */
export const instantiateSha1 = async () => {
    if (cachedSha1.cache !== undefined) {
        return cachedSha1.cache;
    }
    const result = instantiateSha1Bytes(getEmbeddedSha1Binary());
    // eslint-disable-next-line functional/immutable-data, functional/no-expression-statement
    cachedSha1.cache = result;
    return result;
};
