// @bitauth/libauth/build/main/lib/crypto/sha256.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.instantiateSha256 = exports.getEmbeddedSha256Binary = exports.instantiateSha256Bytes = void 0;
const bin_1 = require("../bin/bin");
const format_1 = require("../format/format");
/**
 * The most performant way to instantiate sha256 functionality. To avoid
 * using Node.js or DOM-specific APIs, you can use `instantiateSha256`.
 *
 * @param webassemblyBytes - A buffer containing the sha256 binary.
 */
exports.instantiateSha256Bytes = async (webassemblyBytes) => {
    const wasm = await bin_1.instantiateRustWasm(webassemblyBytes, './sha256', 'sha256', 'sha256_init', 'sha256_update', 'sha256_final');
    return {
        final: wasm.final,
        hash: wasm.hash,
        init: wasm.init,
        update: wasm.update,
    };
};
exports.getEmbeddedSha256Binary = () => format_1.base64ToBin(bin_1.sha256Base64Bytes).buffer;
const cachedSha256 = {};
/**
 * An ultimately-portable (but possibly slower) version of
 * `instantiateSha256Bytes` which does not require the consumer to provide the
 * sha256 binary buffer.
 */
exports.instantiateSha256 = async () => {
    if (cachedSha256.cache !== undefined) {
        return cachedSha256.cache;
    }
    const result = exports.instantiateSha256Bytes(exports.getEmbeddedSha256Binary());
    // eslint-disable-next-line functional/immutable-data, functional/no-expression-statement
    cachedSha256.cache = result;
    return result;
};
