// @bitauth/libauth/build/main/lib/crypto/sha512.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.instantiateSha512 = exports.getEmbeddedSha512Binary = exports.instantiateSha512Bytes = void 0;
const bin_1 = require("../bin/bin");
const format_1 = require("../format/format");
/**
 * The most performant way to instantiate sha512 functionality. To avoid
 * using Node.js or DOM-specific APIs, you can use `instantiateSha512`.
 *
 * @param webassemblyBytes - A buffer containing the sha512 binary.
 */
exports.instantiateSha512Bytes = async (webassemblyBytes) => {
    const wasm = await bin_1.instantiateRustWasm(webassemblyBytes, './sha512', 'sha512', 'sha512_init', 'sha512_update', 'sha512_final');
    return {
        final: wasm.final,
        hash: wasm.hash,
        init: wasm.init,
        update: wasm.update,
    };
};
exports.getEmbeddedSha512Binary = () => format_1.base64ToBin(bin_1.sha512Base64Bytes).buffer;
const cachedSha512 = {};
/**
 * An ultimately-portable (but slower) version of `instantiateSha512Bytes`
 * which does not require the consumer to provide the sha512 binary buffer.
 */
exports.instantiateSha512 = async () => {
    if (cachedSha512.cache !== undefined) {
        return cachedSha512.cache;
    }
    const result = exports.instantiateSha512Bytes(exports.getEmbeddedSha512Binary());
    // eslint-disable-next-line functional/immutable-data, functional/no-expression-statement
    cachedSha512.cache = result;
    return result;
};
